### ISSUE TITLE
ParenthesizedExpression with a JSDoc type-casting comment is not included in the AST

### ISSUE THREAD

**Issue by jens-duttke** [typescript-eslint/typescript-eslint#1943]
**Repro**

```js
/** @type {any} */
const foo = 1;

const bar = /** @type {string} */(foo);
```
* That's just a minimum example to show the problem. I know that it doesn't make sense to cast a number to a string.

```JSON
{
  "rules": {
     "@typescript-eslint/no-unsafe-assignment": "error"
  }
}
```
**Expected Result**

Using `/** @type {string} */`, the type of `foo` is a string. TypeScript handles this correctly. But the rule complains about "Unsafe assignment of an any value", which is not correct.

Seems like the rule ignores the JSDoc type-casting.

**Actual Result**

![no-unsafe-assignment](https://user-images.githubusercontent.com/4883142/80396361-cfcc8780-88b4-11ea-9ef8-d30e947f8570.png)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.29.0` |
| `@typescript-eslint/parser`        | `2.29.0` |
| `TypeScript`                       | `3.9.0-dev.20200419` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `13.6.0` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
Hmmm. I was ready to close this outright, as JSDoc comments are not something we support, however...

This is not actually a bug in the rule. According to the given AST, everything is correctly handled.

It looks like something that's missing support at the parser level.

When we parse the following code `const bar = /** @type {string} */(1 as any);`, we produce the following AST:
```json5
{
  "type": "Program",
  "body": [
    {
      "type": "VariableDeclaration",
      "declarations": [
        {
          "type": "VariableDeclarator",
          "id": {
            "type": "Identifier",
            "name": "bar",
          },
          "init": {
            "type": "TSAsExpression",
            "expression": {
              "type": "Literal",
              "value": 1,
              "raw": "1",
            },
            "typeAnnotation": {
              "type": "TSAnyKeyword",
            },
          },
        }
      ],
      "kind": "const",
    }
  ],
  "sourceType": "module",
}
```

However TS provides the following AST (uninteresting bits removed)
```json5
{
  "kind": 290, // SourceFile
  "statements": [
    {
      "kind": 225, // VariableStatement
      "declarationList": {
        "kind": 243,  // VariableDeclarationList
        "declarations": [
          {
            "kind": 242, // Variable Declaration
            "name": {
              "kind": 75, // Identifier
              "escapedText": "bar"
            },
            "initializer": {
              "kind": 200, // ParenthesizedExpression
              "expression": {
                "kind": 217, // AsExpression
                "expression": {
                  "kind": 8, // NumericLiteral
                  "text": "1"
                },
                "type": {
                  "kind": 125 // AnyKeyword
                }
              }
            }
          }
        ]
      }
    }
  ],
}
```

They look almost identical, apart from one important thing: TypeScript includes an extra `ParenthesizedExpression` node to wrap value.

This is important - because in terms of the types reported by TS - the `AsExpression` / `TSAsExpression` - the type of that thing is `any`.
HOWEVER, the type of the `ParenthesizedExpression` is not `any`, it's `string`!!!

But why? Why does this happen? Why does this work?
Well it TS actually parses JSDoc under the hood (ofc), and it attaches the comments to the nodes where that thinks should be attached to. This isn't publicly exposed, but there is a ***private*** property on the `ParenthesizedExpression`:

```json5
"jsDoc": [
  {
    "kind": 303, // JSDocComment
    "tags": [
      {
        "kind": 319, // JSDocTypeTag
        "tagName": {
          "kind": 75, // Identifier
          "escapedText": "type"
        },
        "typeExpression": {
          "kind": 294, // JSDocTypexpression
          "type": {
            "kind": 143 // StringKeyword
          }
        }
      }
    ]
  }
],
```

Historically, we've always ignored `ParenthesizedExpresson`s because:
- (we thought that) they have no tangible meaning.
- ESTree does not encode parens in the AST. The only times they have syntactical meaning, the AST itself is instead structured to represent that meaning (eg order of operations in math).

In this specific case, the `ParenthesizedExpresson` actually has meaning because it's actually a type assertion (like `TSAsExpression`).
This is something that needs to be fixed at the parser level.

---

Linking this back to the original issue; why does the rule fail?
Well because the parser skips the `ParenthesizedExpression` straight into the inner expression, it means when we ask TS for the type of the initialiser, it returns the type of the `Identifier`, which is OFC `any`.

If we instead output a new node to encapsulate the JSDoc type assertion, then that would instead be what we ask TS for the type, which would instead return `string.

**Comment by Raynos** [comment]
I ran into this myself today when porting from tslint to eslint.

Sounds like this is a bug in the parser.

This is kind of a blocker for using JS since working cast syntax is important. I guess I can disable the eslint rule manually for the line containing the cast.

**Comment by Raynos** [comment]
I see that we handle parens here ( https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/convert.ts#L2266-L2267 ).

We could special case the variable declaration ( https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/convert.ts#L801 ) and assignment expression ( https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/convert.ts#L801 )

Maybe checking a `ParenthesizedExpression` expression for a `jsdoc` field and if it has one returns a `JSDocParen` special AST node ? 

Then handle that `JSDocParen` node through all of the code ...

**Comment by Raynos** [comment]
here are some more examples for test cases

```js
try {
  JSON.parse('foo')
} catch (maybeErr) {
  const err = /** @type {NodeJS.ErrnoException} */ (maybeErr)
  if (err.code !== 'ENOENT') throw err
}
```

```js
try {
  JSON.parse('foo')
} catch (maybeErr) {
  let err
  err = /** @type {NodeJS.ErrnoException} */ (maybeErr)
  if (err.code !== 'ENOENT') throw err
}
```

```js
if (groupsStr) {
  const groupsInfo = /** @type {{
    groups: string[]
  }} */ (JSON.parse(groupsStr))
  this.populateGroups(groupsInfo.groups)
}
```

**Comment by bradzacher** [comment]
my thinking for this was to add special handling for just this case.
Though it's important to note that "just this case" is not a trivial amount of work.

when we parse the `ParenthesizedExpression`, we need to first check if:
- the first child is a node of kind `SyntaxKind.JSDocComment`.
- `JSDocComment.tags` contains a node of kind `SyntaxKind.JSDocTypeTag`.

If this is false, ignore the `ParenthesizedExpression`.
If that is true, emit a new ast node.

Now I'm not entirely sure what the AST node should look like...
I think that this is the best path forward.

We treat this as a special, one-off case, and emit an AST node like this:
```ts
interface TSJSDocTypeAssertion extends BaseNode {
  type: AST_NODE_TYPES.TSJSDocTypeAssertion;
  expression: Expression;
  typeAnnotation: JSDocTypeNode;
}
```


Note the `JSDocTypeNode` - we would have to ensure we're handling all of the special JSDoc type cases.
Having a look at the cases that TS supports out of the box, it would look something like:
```ts
type JSDocTypeNode =
  | TypeNode // existing AST nodes
  | TSJSDocAllType
  | TSJSDocFunctionType
  | TSJSDocNonNullableType
  | TSJSDocNullableType
  | TSJSDocOptionalType
  | TSJSDocUnknownType
  | TSJSDocVariadicType;
// https://github.com/microsoft/TypeScript/blob/5b0194b3114eca8ebe32695bc0603d6a6bf01b7d/lib/typescript.d.ts#L1588-L1616
```

Testing with real code:
```ts
// already supported by our parser as it reuses existing ts ast nodes
const keyword = /** @type {string} */ (value)
const typeReference = /** @type {reference} */ (value)
const typeReferenceQualifiedName = /** @type {reference.prop} */ (value)
const typeReferenceWithGeneric1 = /** @type {Array<string>} */ (value)
const typeReferenceWithGeneric2 = /** @type {Array.<string>} */ (value) // NOTE - IDK if we want to treat this differently to the normal generic type ref.
const specialRecordType = /** @type {Object.<string, number>} */ (value) // NOTE - TS parses this as a TypeReference to Object, but the resolved type is correctly set to { [x: string]: number; }
const arrayType = /** @type {string[]} */ (value)
const union = /** @type {string | number} */ (value)
const parenthesizedType = /** @type {(string | number)} */ (value)
const typeLiteral = /** @type {{ A: string }} */ (value)
const importType = /** @type {import('typescript').SyntaxKind} */ (value)


// needs new nodes
const TSJSDocAllType = /** @type {*} */ (value)
const TSJSDocFunctionType = /** @type {function (a: string): number} */ (value)
const TSJSDocNullableType = /** @type {?string} */ (value)
const TSJSDocNonNullableType = /** @type {!string} */ (value)
const TSJSDocOptionalType = /** @type {string=} */ (value)
const TSJSDocUnknownType = /** @type {?} */ (value)
const TSJSDocVariadicType = /** @type {...string} */ (value)
```
[repl](https://ts-ast-viewer.com/#code/FAEwpgxgNghgTmABBA9gOwM4BdEDcZQCuYAXIjGgJ4DcwwA9PeVAjCJYhoQA7cpxYwIchkQBLHAkIYwosAA8x2MWgDmnSmiwx5wVJhwBrMJQDu-YQF5E9AFS3EAASyVuSAN7Y4K1QF9EtkwAFPhEYACUeujYiC5uAEpgAGZgCGgQSNZ2Ds6uHggpaRn+gYghBMSR+jFxYImFYOlgAIqEBGJJYkIAcjAAtpk29k61iO4FqY0ZAHTccCjcJcGhlVEGsXn1k00A6hIAFgDijaliEACMiFnDuW5jAIJwcDCUADxePgB8S2UrEWs1TbJbYZPZYI4nbwQABMVyGOVG7kez0o03eWG8am+AWWFQiNiY3QA8gAVACiiAAtIgAJIAEQA0uIkohTEhTBQcFgULFWFz9kpECAOg0tFAONzYvskGh+H0CIhVJCzhs7gVpgCcBg3BAxAREqg4CASXk4dkRqb3ESAEYAK0gWDRHzUABpEGhCH1ralsaVymFwgTEMTyVTECSAMqIbjwGSicGCmCiGDhoGijKxHk2+0QLBu62EflIBAYFBQXBCVVIQWGhC58WcMBcnnuRAAbXkZGdqgAumQPV7UtREL5NeQni8TXdrgjLd22z2fv7VtUcIQ0GJ0Gaboju4gAD7uz3euBLv5VaI4GNpcGyMQALyEU8G5tuHiCe8PA5P4TPeIv6y1AAMhIqQKjOFp3O4rb3F2GI+COf4BmOYh9HwAjPtus5Qah6FYEEADktQYBA3jcFgBHhNMEaaNo8gMioIBIasdCMO6YBCKIaBgKY7ooOAGBjpGABSEZ0igED3FAUCYRBb5jLYzH-Ku4YRqJ4kQN0hDSTA1pQGAsnwpBHgAPzdkpAExCJYkSd06BaTpekGaacmIgAhOZOK-P+QlqTZEAAGrwHqwoQIZr6ItMUWeX656+epElEuRm5oAQ4U7nO8FqJYFnxf5ACqaCGLKphoOl2GmblKnWRpABi665il5XGWMSQNVgKVlDAcGYqo4T9seqS5XQDCMGx5IRiSNLdIciA7AAEmS8QUgAwkSACy61kt0kaIPcy2IAACkB9wrWSdJjqgfQDFoogzsAASxLIWDnA9FVjJ5gjYNCb24shI1jYDQPAyDoNg2No3BqSiARvlh2HUS8TknSiAAEIAJqqTDRJzat9zdLShzEgdNIkgSo3g5TVPg6xTC7cRUqyEgSaILa0g4NmDpurKOBIiQ35DWOtoYBpnO5s+lz3Y98lWnaDr+H0lDZm9LXuALp6IIr2bTPcv3echKnC6LYtYM+sJS+9ss5lgCtK3aKuOHMCypC4Yzq-4utecu-yQwlEDuv0SAxuCyZpERnA8HhlZJPw0axgh1oShgLpCyLtmB4dMDgpLRkyxMRRgAAfk7ixe3Fhtp5pGdZ-s5u54i+dTGAADEJfDRT1Od133c973kM7EgIDoOH3GVpKbDCAo3BQGcEgR7w-A4DHcBSkobos2yBEIIg6ANteGAIbe4hoDP3EiDIAidYwHe97fd-353AN01GMCFig8odRABB77GsiMzvctcxViFMkFQEhOoUGEKRMAWc-4plGKYAUEB9jIAoIgb0iBG5NBAAwewDtajgBZIdeYbhL6yHmkmfY+DLQoEATbB2JcXYcE8FlPw4hhDUjQfSDU0tGECGYZ5VKAwwyUBQIQFeQiwA8JyHw12atBoaxgEqERYiV5KKkb9OgKkS53XrpaYhzsyEYAoRgfYSxvbhFoJDXar9uQfzON-Dg+8-5Hy-tJa0MAICGGAYQsBHUtyQOQHyOBwDEFnBQV-NA6DizAgLjg7IDs3FQA8V4zWlAVrfxSYYBh8B+huwUf4eQYYorSPoFoy8qD3GeO8a5S0isMlVK8eY88tAgA)



In future if we want to add full support for JSDoc comments, we can simply extend this node to include a comment property:
```ts
interface TSJSDocTypeAssertion extends BaseNode {
  type: AST_NODE_TYPES.TSJSDocTypeAssertion;
  expression: Expression;
  typeAnnotation: TypeNode | JSDocTypeNode;
  comment: TSJSDocComment;
}

interface TSJSDocComment extends BaseNode {
  type: AST_NODE_TYPES.TSJSDocComment;
  tags: TSJSDocTag[];
  comment?: string;
}

type TSJDocTag =
  | TSJSDocUnknownTag
  | TSJSDocAugmentsTag
  | TSJSDocAuthorTag
  | TSJSDocClassTag
  | TSJSDocPublicTag
  | TSJSDocPrivateTag
  | TSJSDocProtectedTag
  | TSJSDocReadonlyTag
  | TSJSDocEnumTag
  | TSJSDocThisTag
  | TSJSDocTemplateTag
  | TSJSDocReturnTag
  | TSJSDocTypeTag
  | TSJSDocTypedefTag
  | TSJSDocCallbackTag
  | TSJSDocPropertyTag
  | TSJSDocParameterTag;
// https://github.com/microsoft/TypeScript/blob/5b0194b3114eca8ebe32695bc0603d6a6bf01b7d/lib/typescript.d.ts#L1629-L1727
```

**Comment by Raynos** [comment]
Thanks for that detailed write up. will be super uesful.

**Comment by Josh-Cena** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/1682
